### DISCUSSION TITLE
[Questions] RVC on ROCm 6.4

### DISCUSSION THREAD

**Discussion by robertrosenbusch** [robertrosenbusch/gfx803_rocm#27] [Ideas]
@lsunay : Welcome! I have some questions to integrate/migrate the RVC-ROCm 5.4 Stuff to ROCm 6.4.

Thanks a lot of your detailed README.md and all the Stuff from your PR! I am definitely not very close on your RVC Project, but i am curious and i wanna support it with your help :P

I got some questions, while i try to rebuild your ROCm 5.4 RVC-Project on your README.md. **All (python) tests failed. I am sorry.**

`RUN echo "** Verifying PyTorch, TorchVision, TorchAudio Installation **" && \
    python3 -c "import sys; import platform; import torch; import torchvision; import torchaudio; \
                print(f'System Python Version (sys.version): {sys.version.split()[0]}'); \
                print(f'Platform Python Version (platform.python_version): {platform.python_version()}'); \
                print(f'Torch Version: {torch.__version__}'); \
                print(f'TorchVision Version: {torchvision.__version__}'); \
                print(f'TorchAudio Version: {torchaudio.__version__}'); \
                print(f'ROCm available: {torch.cuda.is_available()}'); \
                print(f'PyTorch built with ROCm/HIP: {torch.version.hip if hasattr(torch.version, \"hip\") else \"HIP not available or not a ROCm build\"}')"`
                
But on my own [PyTorch 2.6 Container ](https://github.com/robertrosenbusch/gfx803_rocm/blob/main/Dockerfile_rocm64_pytorch) your short Script doesnt failed.

Lets try out on this ROCm6.4/PyTorch 2.6 Container. I am not really excited to made or debug your RVS-Project on ROCm 5.4. There is a lot of chinese stuff inside :P Srry.

**Comment by lsunay** [comment]
Hi Robert, 

Thanks for the detailed feedback and for trying to set up the ROCm 5.4 environment based on my README.md. I also appreciate your willingness to collaborate on potentially migrating RVC to ROCm 6.4. 

I'm quite surprised to hear that all Python tests failed when you tried to rebuild the ROCm 5.4 RVC project. The dependencies, including PyTorch 2.0.1, TorchVision 0.15.2, and TorchAudio 2.0.2 with Python 3.10, are specifically pinned for that ROCm 5.4.2 environment (rocm/dev-ubuntu-22.04:5.4.2-complete -py3.10.) , and the verification script you shared should ideally pass. 

To help me understand what might have gone wrong with your ROCm 5.4 setup attempt, could you please share a few things? 

     The Dockerfile you used for the ROCm 5.4 attempt: Did you use the one from my PR/repository directly, or did you adapt it? Sharing the exact Dockerfile would be very helpful.
     The full build logs and the run logs (especially the output showing the Python test failures) from your ROCm 5.4 attempt.
     

It's good to know that the verification script works correctly in your own PyTorch 2.6 container on ROCm 6.4. This confirms your base environment is healthy for that PyTorch version.
You suggested trying RVC on this ROCm 6.4/PyTorch 2.6 container. To evaluate this path and understand your setup better: 

     Could you also share the Dockerfile for your working ROCm 6.4 / PyTorch 2.6 container? This would give me insight into the base image and how PyTorch 2.6 is configured.
     

As I mentioned previously, the main challenge with RVC on newer ROCm versions like 6.4 is RVC's own dependency on older libraries (like PyTorch 2.0.1). Even if your PyTorch 2.6 environment is solid, RVC itself is coded to expect the APIs and behaviors of PyTorch 2.0.1. Trying to run it directly on PyTorch 2.6 would likely lead to different kinds of errors (API mismatches, etc.) unless RVC's code is updated to be compatible with PyTorch 2.6. 

Regarding the "Chinese stuff" in the project, I understand that can sometimes be a hurdle with documentation or comments. 

Once I can review your Dockerfiles and logs for the ROCm 5.4 attempt, we might get a clearer picture of the initial issue. It would also be good to see your ROCm 6.4 Dockerfile to understand the environment you're proposing. 

Looking forward to your reply! 

Best regards, 

Note:
Regarding the possibility of compiling RVC for ROCm 6.4, I wanted to elaborate on the challenges I foresee, largely due to the RVC project not having been updated for about two years. This has led to its dependencies becoming quite outdated for newer environments like ROCm 6.4. 

The specific dependencies from RVC's requirements.txt (or its setup) are: 

     PyTorch GIT: v2.0.1
     TorchVision GIT: v0.15.2
     TorchAudio GIT: v2.0.2
     Python: 3.10
     

This specific combination (PyTorch 2.0.1, etc., with Python 3.10) is precisely why, for my previous work (and the PR you mentioned for ROCm 5.4), I opted for a base image like rocm/pytorch:rocm5.4.2-runtime-ubuntu20.04-py3.10. This older ROCm version and its corresponding PyTorch build are more aligned with these RVC dependencies. 

When I, similar to what you might have experienced with your attempts, tried to compile RVC for ROCm 6.4, I couldn't get past these dependency issues. ROCm 6.4 generally requires much newer versions of PyTorch (e.g., 2.2+ or later, specifically built for ROCm 6.x) for proper compatibility and to leverage its features. The PyTorch 2.0.1 ecosystem is simply too old for a smooth build and runtime experience on ROCm 6.4, especially with Python 3.10.

**Comment by robertrosenbusch** [comment]
Hi @lsunay : I got this one from your PR on RVC into building on to my GIT ...  
[rvc_rocm54.txt](https://github.com/user-attachments/files/20494513/rvc_rocm54.txt) 

But relax, we are able to do it on ROCm 6.4 succesfully too. Depends on you  :+1: 

---
Cheers, Robert Rosenbusch

**Reply by lsunay** [reply]
"Hi Robert,

Thanks for your patience as we've been discussing this.

You are correct that the build logs for my Stage 1 (base ROCm 5.4) image show these warnings:
*   `UserWarning: Failed to load image Python extension...` (for TorchVision)
*   `UserWarning: No audio backend is available.` (for TorchAudio)
*   And `ROCm available: False` during that build-time check.

I can confirm these warnings also appear in my own successful builds of the Stage 1 base image.

However, the key point is that these warnings in the **Stage 1 base image build process do not prevent the final RVC application from working correctly for voice cloning.**

Here's why this setup still works for me:

1.  **Multi-Stage Build:** My RVC setup uses a two-stage Docker build.
    *   **Stage 1 (`Dockerfile.base_rocm5.4_source_compile`)** compiles PyTorch, TorchVision, and TorchAudio. The warnings you see are from the verification script at the end of *this* stage.
    *   **Stage 2 (`Dockerfile.rvc_original`)** takes this base image and then installs the RVC WebUI application and its specific Python dependencies into a virtual environment.

2.  **RVC's Dependencies in Stage 2:** It seems that the Python packages installed by RVC in Stage 2 (like `soundfile`, `pydub`, `librosa`, etc., listed in its `requirements.txt` or installed manually in my Stage 2 Dockerfile) provide the necessary audio handling capabilities at the Python level, effectively working around the "No audio backend available" warning seen during the base PyTorch/TorchAudio compilation. For instance, the `soundfile` Python package can use the `libsndfile1` system library (which *is* present in my base image after the `purge` of `-dev` packages) to handle audio.

3.  **TorchVision Not Critical for RVC Audio:** The TorchVision warning is ignorable because RVC's core voice conversion functionality does not rely on TorchVision's image I/O capabilities. This was a conscious decision to keep the base image leaner.

4.  **`ROCm available: False` during Build:** This is common during Docker `RUN` commands as the build environment might not have full GPU access. The critical part is that PyTorch *is* built with ROCm/HIP support, and ROCm becomes available when the final container is run with correct Docker flags (e.g., `--device=/dev/kfd --device=/dev/dri`).

**So, if you are trying to replicate my setup:**

*   The warnings you see when building the Stage 1 base image are "expected" in the sense that they don't stop the final RVC application (built in Stage 2) from working for its intended audio purpose.
*   The "failures" you mentioned might be a misinterpretation of these build-time warnings from Stage 1. The true test is whether the RVC application, once fully built through Stage 2, can perform voice cloning.

Could it be that the "tests" you mentioned failing were these verification script outputs from the Stage 1 build? If so, I'd suggest proceeding to build Stage 2 on top of that base image and then testing the RVC application itself.

If you are still encountering issues with the RVC application *after* a complete two-stage build, then we would need to look at the logs from Stage 2 or from running the RVC application.

Let me know if this clarifies things from my end. My setup produces those Stage 1 warnings but results in a functional RVC for audio.

Best regards,

> Note:
> "Additionally, I wanted to note that I've provided logs from five different attempts for each main compilation stage for your review. The logs from the very last successful builds are build_rocm542_v4.log (for the base ROCm 5.4.2 image) and build_rvc14.log (for the RVC application layer). Since I utilized Docker's build cache, the earlier logs are included to offer visibility into the specific details of intermediate steps, should you need to trace them. I mention this to highlight that a considerable amount of time and effort was dedicated to achieving a fully successful and working build.

---

**Comment by lsunay** [comment]
---

Hi Robert,

Following up on our discussion about potentially moving RVC to ROCm 6.4, I wanted to clearly outline the main reasons why I believe compiling the current RVC project directly for a ROCm 6.4 environment would be very challenging, if not currently infeasible.

The core issue stems from the fact that the **RVC project itself has not been significantly updated for approximately two years.** This has led to its dependencies being pinned to older versions of key libraries, particularly PyTorch and its ecosystem.

Here's a breakdown of the challenges:

1.  **Outdated PyTorch Dependency:**
    *   The RVC project, as per its original requirements (and my successful ROCm 5.4 build), is built against **PyTorch v2.0.1**, TorchVision v0.15.2, and TorchAudio v2.0.2.
    *   ROCm 6.4, on the other hand, is designed to work best with much newer versions of PyTorch (e.g., PyTorch 2.2+ or later, specifically built with ROCm 6.x support). These newer PyTorch versions are necessary to leverage the features, bug fixes, and performance improvements of the ROCm 6.4 stack.

2.  **API and ABI Incompatibilities:**
    *   Attempting to compile or run RVC (which expects PyTorch 2.0.1 APIs) with a PyTorch version suitable for ROCm 6.4 (e.g., PyTorch 2.2+) would almost certainly lead to:
        *   **Compilation errors:** RVC's C++/CUDA extensions (if any, or those within its direct dependencies like `fairseq`) might fail to compile against newer PyTorch headers.
        *   **Runtime errors:** Significant API differences between PyTorch 2.0.1 and PyTorch 2.2+ would cause crashes or incorrect behavior in RVC's Python code. Functions may have been renamed, their signatures changed, or their behavior altered.

3.  **Dependency Hell with Associated Libraries:**
    *   It's not just PyTorch. RVC depends on specific versions of other libraries like `fairseq`, `numpy`, `numba`, `llvmlite`, etc., which were compatible with PyTorch 2.0.1.
    *   Trying to update *only* PyTorch to a ROCm 6.4-compatible version would likely break these other dependencies, requiring a cascade of updates throughout RVC's dependency tree. This is a non-trivial task and could involve significant code changes within RVC or its dependencies if newer compatible versions don't exist or behave differently.

4.  **Build System and Toolchain Challenges:**
    *   Compiling older PyTorch versions (like 2.0.1) directly against a newer ROCm stack (6.4) is often problematic. The build systems and compiler flags expected by older PyTorch might not align with what ROCm 6.4 provides or requires, leading to build failures.
    *   My ROCm 5.4 Dockerfile specifically uses `rocm/dev-ubuntu-22.04:5.4.2-complete` as a base and pins Python to 3.10 because that combination was known to work for PyTorch 2.0.1. Replicating this for ROCm 6.4 with PyTorch 2.0.1 is unlikely to succeed smoothly.

5.  **Python Version Compatibility:**
    *   While RVC uses Python 3.10, which is generally fine, newer PyTorch builds for newer ROCm versions might also start recommending or requiring newer Python sub-versions, potentially adding another layer of complexity if RVC has strict Python 3.10 sub-version dependencies in its other packages.

In essence, porting RVC to ROCm 6.4 isn't just a matter of pointing it to a new ROCm version. It would likely require a **significant engineering effort to update the RVC codebase and/or its core dependencies** to be compatible with the newer PyTorch ecosystem that ROCm 6.4 expects.

Given that RVC hasn't seen active development to keep up with these library changes, a direct compilation on ROCm 6.4 with its current dependencies is, in my assessment, not a straightforward path and is prone to failure.

This is why my efforts were focused on getting it to work on ROCm 5.4, which aligns better with RVC's last known stable dependency set.

I'm open to hearing your thoughts or if you have a different perspective on tackling these dependency challenges for ROCm 6.4.

Best regards,

Levent

> Note: 
> With RVC, you can train a model using an approximately 8-minute audio sample, and then it can clone another audio file to perfectly match that sample voice. I found it gave quite satisfactory results even with a brief test. There's a YouTube video below that demonstrates this, which you might find interesting. My main reason for choosing RVC for voice cloning, despite some of the build complexities we've discussed, was its proven success as an application. 
> You can see an example here: https://youtu.be/g4td-n8-QZk?si=b6grlvRZZ4SYV784

---

**Reply by robertrosenbusch** [reply]
@lsunay : Hi Levent and welcome back! I guess we are able to "improve" your RVC Thing.

> * The RVC project, as per its original requirements (and my successful ROCm 5.4 build), is built against **PyTorch v2.0.1**, TorchVision v0.15.2, and TorchAudio v2.0.2.
> 
> * ROCm 6.4, on the other hand, is designed to work best with much newer versions of PyTorch (e.g., PyTorch 2.2+ or later, specifically built with ROCm 6.x support). These newer PyTorch versions are necessary to leverage the features, bug fixes, and performance improvements of the ROCm 6.4 stack.

Mmmhhh, just a hint... there is a [rocm6.1.2_ubuntu20.04_py3.9_pytorch_release-2.0.1](https://hub.docker.com/layers/rocm/pytorch/rocm6.1.2_ubuntu20.04_py3.9_pytorch_release-2.0.1/images/sha256-c06d6389d345cdef1b28a47d27174344742a6bccd36165515cd88dccb2166224) which might be will full fill your dreams :P Cause i measured +20% performance speed up ROCm 5.5<->5.7 and another +15%performance speed up on ROCm 5.7<-> 6.3 on ComfyUI and on PyTorch
. i try a little around a long time ago and remeber it. Nevertheless... When AMD official published the PyTorch 2.0.1 on Python 3.9 on ROCm 6.1.2...

**Where is the Prob to** compile PyTorch 2.0.1 on ROCm 6.4 under Buntu 24.04 with Python 3.10 into Env and all the necessary Stuff for your RVC?  And then to install the RVC lib things?

Maybe i am thinking a loud :P I am sorry ^.^

---
Cheers, Robert Rosenbusch

**Reply by lsunay** [reply]
Hi Robert (@robertrosenbusch), 

Thanks for the heads-up and that link – that's an awesome find! You're absolutely right, the `rocm/pytorch:rocm6.1.2_ubuntu20.04_py3.9_pytorch_release-2.0.1` image looks very promising. If PyTorch 2.0.1 is officially supported and built by AMD for ROCm 6.1.2 with Python 3.9, it certainly suggests that compiling PyTorch 2.0.1 for that specific ROCm version and Python 3.9 is feasible, especially since the original RVC project also runs fine on Python 3.9. This could indeed simplify things. 

My main concerns, and the "Prob" I was referring to, specifically relate to the scenario of attempting to compile the older PyTorch 2.0.1 directly against ROCm 6.4 (especially on a newer OS like Ubuntu 24.04 with Python 3.10). This is where I anticipate challenges like: 

1. **Build System Incompatibilities:** The PyTorch 2.0.1 build system might expect an older toolchain that doesn't align perfectly with what a much newer ROCm 6.4 environment provides.
2. **ROCm API/ABI Nuances:** Potential subtle differences in newer ROCm versions that an older PyTorch build might not handle.
3. **Dependency Hell:** Satisfying PyTorch 2.0.1's dependencies on a much newer OS and Python version can be complex. The official AMD image you found likely has these carefully managed for its specific (ROCm 6.1.2, Py3.9, Ubuntu 20.04) combination.

 I believe compiling PyTorch 2.0.1 on ROCm 6.1.2 with Python 3.9 (as suggested by the official AMD image) is likely achievable. However, when considering the effort vs. reward for pushing that specific PyTorch 2.0.1 build to ROCm 6.4 with Python 3.10, especially for the RVC project which hasn't been updated approximately 2 or 3 years, it feels like the endeavor might not be worth battling through the potential "dependency hell" and other compilation issues. 

Given that the RVC repository is somewhat outdated, instead of investing significant effort to make its old dependencies work on the very latest ROCm, it seems more productive to focus on more actively maintained and up-to-date alternatives. These newer projects are more likely to be compatible with, or easier to adapt to, modern PyTorch versions that are already well-supported and compiled for ROCm 6.4 (like the PyTorch 2.1+ or 2.x versions you've mentioned working with in your ComfyUI tests). 

For instance, one such project I'm currently looking into is Chatterbox by Resemble AI ([https://github.com/resemble-ai/chatterbox ](https://github.com/resemble-ai/chatterbox)). It's a more modern application (demo: [https://huggingface.co/spaces/ResembleAI/Chatterbox ](https://huggingface.co/spaces/ResembleAI/Chatterbox)). Although it currently only supports English, adapting a project like this, which is likely using more current PyTorch practices, to a pre-existing ROCm 6.4 + PyTorch 2.6 environment seems like a more efficient path forward. 

Thanks again for pointing out that specific image and for the insightful discussion! It really helps clarify the different viable paths. 

Best regards,
Levent Sunay

**Comment by chboishabba** [comment]
@robertrosenbusch @lsunay 

```
This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)
/Whisper-WebUI/venv/lib/python3.12/site-packages/torch/cuda/memory.py:391: FutureWarning: torch.cuda.reset_max_memory_allocated now calls torch.cuda.reset_peak_memory_stats, which resets /all/ peak memory stats.
  warnings.warn(
HW Exception by GPU node-1 (Agent handle: 0x18c30940) reason :GPU Hang
/Whisper-WebUI/whisp.sh: line 2:     7 Aborted                 (core dumped) ./venv/bin/python3 app.py --whisper_type faster-whisper --share --server_name 0.0.0.0 --server_port 7860
```
